modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java

```diff
@@ -0,0 +1,382 @@
+        Path serverDataPath = toStop.getServerDataPath(raftGroupId());
+
+        // Get the path to that node's RocksDB key-value storage
+        Path dbPath = getListenerPersistencePath(getListener(toStop, raftGroupId()));
+
+        int stopIdx = servers.indexOf(toStop);
+
+        // Remove that node from the list of servers
+        servers.remove(stopIdx);
+
+        // Shutdown that node
+        toStop.stop();
+
+        // Create a snapshot of the raft group
+        service.snapshot(service.leader()).get();
+
+        doAfterStop(service);
+
+        // Create another raft snapshot
+        service.snapshot(service.leader()).get();
+
+        if (testData.deleteFolder) {
+            // Delete a stopped node's raft directory and key-value storage directory
+            // to check if snapshot could be restored by the restarted node
+            IgniteUtils.deleteIfExists(dbPath);
+            IgniteUtils.deleteIfExists(serverDataPath);
+        }
+
+        if (testData.interactAfterSnapshot) {
+            // Interact with the raft group after the second snapshot to check if the restarted node would see these
+            // interactions after restoring a snapshot and raft logs
+            doAfterSnapshot(service);
+        }
+
+        // Restart the node
+        JRaftServerImpl restarted = startServer(stopIdx);
+
+        assertTrue(waitForTopology(cluster.get(0), servers.size(), 3_000));
+
+        BooleanSupplier closure = snapshotCheckClosure(restarted, testData.interactAfterSnapshot);
+
+        boolean success = waitForCondition(closure, 10_000);
+
+        assertTrue(success);
+    }
+
+    /**
+     * Interacts with raft group before the follower is stopped.
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doBeforeStop(RaftGroupService service) throws Exception;
+
+    /**
+     * Interacts with raft group after the follower is stopped.
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doAfterStop(RaftGroupService service) throws Exception;
+
+    /**
+     * Interacts with raft group after the leader has captured a snapshot.
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doAfterSnapshot(RaftGroupService service) throws Exception;
+
+    /**
+     * Creates a closure that will be executed periodically to check if the snapshot and (conditionally on the
+     * {@link TestData#interactAfterSnapshot}) the raft log was successfully restored by the follower node.
+     *
+     * @param restarted Restarted follower node.
+     * @param interactedAfterSnapshot {@code true} whether raft group was interacted with after the snapshot operation.
+     * @return Closure.
+     */
+    public abstract BooleanSupplier snapshotCheckClosure(JRaftServerImpl restarted, boolean interactedAfterSnapshot);
```

Comment:
Actually I think that all these abstract methods are a sign of over-engineering. Do we really need an abstract class for these two tests? I think it is easier to copy-paste the test itself....

Reply:
Tests are basically the same, any change in the raft group API would require changes in two places

Reply:
Yes, but all these abstract methods show that they are not really the same, only the overall test scenario

Reply:
I think raft set up, stop and restart operations are too complicated to copy-paste them. I don't see how abstract tests can hurt us